sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala

```diff
@@ -0,0 +1,179 @@
+   * event backend.
+   */
+  def start(system: ActorSystem, host: String, port: Int): EventingTestKit = {
+
+    // Create service handlers
+    val service = new EventingTestServiceImpl(system, host, port)
+    val handler: HttpRequest => Future[HttpResponse] =
+      EventingTestKitServiceHandler(new service.ServiceImpl)(system)
+
+    val binding = Await.result(
+      Http(system)
+        .newServerAt(host, port)
+        .bind(handler),
+      10.seconds)
+    // to allow binding to 0
+    val actualPort = binding.localAddress.getPort
+    system.log.info("Eventing testkit grpc-backend started at {}:{}", host, actualPort)
+    service.port = actualPort
+    service
+  }
+
+}
+
+/**
+ * Implements the EventingTestKit protocol defined in protocols/testkit/src/main/protobuf/eventing_test_backend.proto
+ */
+private[testkit] final class EventingTestServiceImpl(system: ActorSystem, val host: String, var port: Int)
+    extends EventingTestKit {
+  import EventingTestKitImpl._
+
+  override def getHost(): String = host
+  override def getPort(): Integer = port
+
+  private val log = LoggerFactory.getLogger(classOf[EventingTestKit])
+  private implicit val sys = system
+  private implicit val ec = sys.dispatcher
+
+  private val eventDestinationProbe = new ConcurrentHashMap[EventDestination, TestProbe]()
+
+  override def getTopic(topic: String): Topic = {
+    val dest = EventDestination.defaultInstance.withTopic(topic)
+    eventDestinationProbe.putIfAbsent(dest, TestProbe())
+    val probe = eventDestinationProbe.get(dest)
+    new TopicImpl(probe)
+  }
+
+  final class ServiceImpl extends EventingTestKitService {
+    override def emitSingle(in: EmitSingleCommand): Future[EmitSingleResult] = {
+      log.info("emitSingle request: [{}]", in)
+
+      in.destination.foreach(dest => {
+        eventDestinationProbe.putIfAbsent(dest, TestProbe())
+        eventDestinationProbe.get(dest).ref ! in
+      })
+
+      if (in.destination.isEmpty) {
+        log.warn("Received a message without destination, ignoring. {}", in)
+      }
+      Future.successful(EmitSingleResult())
+    }
+
+    override def runSource(in: Source[RunSourceCommand, NotUsed]): Source[SourceElem, NotUsed] = ???
+
+    override def eventStreamOut(in: Source[EventStreamOutCommand, NotUsed]): Source[EventStreamOutResult, NotUsed] = ???
+  }
+}
+
+class TopicImpl(private val probe: TestProbe) extends Topic {
+  @ApiMayChange
+  override def expectNext(): TestKitMessage[ByteString] = {
+    val msg = probe.expectMsgType[EmitSingleCommand]
+
+    val metadata = new MetadataImpl(msg.getMessage.metadata.getOrElse(Metadata.defaultInstance).entries)
+    new TestKitMessageImpl(msg.getMessage.payload, metadata)
+  }
+
+  @ApiMayChange
+  override def expectAll(): util.Collection[TestKitMessage[ByteString]] = {
+    probe
+      .receiveWhile(3 seconds) { case m: EmitSingleCommand =>
```

Comment:
this one will be tricky to use, how about exposing similar semantics to be clear that about the outcome?
```
override def receiveWhile(duration): util.Collection[TestKitMessage[ByteString]] = {
```

Reply:
Yeah, at least expose the timeout for sure. I'm not quite settled on the API namings. In any case, `expectAll` is not great indeed.. Let me do another round of clean up and experimentation and see where we land..